### PR TITLE
[2616] Add a temporary edge during connection

### LINKED ISSUE #2616: With React Flow the connection line is lost during relation type choice
If there are two or more connectorTools on an edge creation, a popup is display to select the wanted connection.
But at this point, we lost the connection line betwenn the two nodes, which can be disturbing for the user.

In sprotty with the magic connector we don't have this behavior.

### Sprotty behavior

edge_connector_sprotty

### React flow behavior

edge_connector_react_flow

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: [#2616](https://github.com/eclipse-sirius/sirius-web/issues/2616)

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Review by gcoutable — CHANGES_REQUESTED** [review]
I figured out at least one small bug, if the edge creation is cancelled, the selected creation handle seems to be selected still.
Another point that could be improved: the temporary connection line may not start from the creation handle the end-user has clicked.

https://github.com/eclipse-sirius/sirius-web/assets/7371042/b84ea5d7-87a5-44af-a27e-6c256c2ca1b9

**Comment by mcharfadi** [comment]
I fixed the issue with the highlight of the handle.
Concerning the temporary connection line that is not starting from the handle clicked, I think it's more convenient to have it start from the side that will be used to render the futur edge.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx [hunk 1]
```diff
@@ -184,6 +186,44 @@ export const ConnectorContextualMenu = ({}: ConnectorContextualMenuProps) => {
...
+        updatable: false,
+        zIndex: 2002,
+      };
+      reactFlowInstance.setEdges((oldEdges: Edge<EdgeData>[]) => [...oldEdges, edge]);
+    }
+  };
+
+  const removeTempConnectionLine = () => {
+    reactFlowInstance.setEdges(() => reactFlowInstance.getEdges().filter((item) => !item.id.includes('temp')));
+  };
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: Can you extract that into a hook...

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx [hunk 2]
```diff
@@ -210,4 +251,11 @@ export const ConnectorContextualMenu = ({}: ConnectorContextualMenuProps) => {
       ))}
     </Menu>
   );
-};
+});
+
+export const ConnectorContextualMenu = memo(({}: ConnectorContextualMenuProps) => {
+  const connectionNodeId = useStore((state) => state.connectionNodeId);
+  const isConnecting = !!connectionNodeId;
+  const { connection } = useConnector();
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: ... include something like `connectionInProgress()`  which does `!!connectionNodeId || !!connection`.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx [hunk 3]
```diff
@@ -210,4 +251,11 @@ export const ConnectorContextualMenu = ({}: ConnectorContextualMenuProps) => {
...
+  return isConnecting || !!connection ? <ConnectorContextualMenuComponent /> : null;
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: ... Thus, replace this line by `connectionInProgress ? <ConnectorContextualMenuComponent /> : null;`

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionCreationHandles.tsx [hunk 4]
```diff
@@ -105,6 +105,18 @@ export const ConnectionCreationHandles = ({ nodeId }: ConnectionCreationHandlesP
...
+  const isConnecting = !!connectionNodeId;
+  const { connection } = useConnector();
+  useEffect(() => {
+    if (!isConnecting && !connection) {
+      setState((prevState) => ({
+        ...prevState,
+        isMouseDown: null,
+      }));
+    }
+  }, [isConnecting, connection]);
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionCreationHandles.tsx]: ...Then adapt this `useEffect` to rely on `connectionInProgress()` from the hook.

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx (hunk 1 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: done

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx (hunk 2 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: done

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx (hunk 3 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/connector/ConnectorContextualMenu.tsx]: done

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionCreationHandles.tsx (hunk 4 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionCreationHandles.tsx]: done

**Review by gcoutable — APPROVED** [review]

**Pull request merged**